### PR TITLE
Refactor Alert module #1694

### PR DESCRIPTION
**Pull request by elonlo**
## What is the purpose of the pull request

Refactor Alert module 
*Step 1: define the Sender interface, use JAVA SPI to decouple the sender logic, and support the registration of alarm plugins.*

## Brief change log

  - *Add powermock to alert pom.xml*
  - *Add sender package to module*

## Verify this pull request

This change added tests and can be verified as follows:
  - *Added EmailSenderTest to verify the change.*
  - *Added SenderManagerTest to verify the change.*
  - *Added ShortMessageServiceSenderTest to verify the change.*
  - *Added AlertExceptionTest to verify the change.*
  - *Manually verified the change by testing locally.*

### CODE REVIEW

File: dolphinscheduler-alert/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java [hunk 1]
```diff
@@ -73,4 +76,37 @@ public static void main(String[] args){
         alertServer.start();
     }
 
+    /**
+     * send alert message
+     * @param alerts waiting list
+     */
+    private void send(List<Alert> alerts) {
```
Comment (Technoboy-) [inline dolphinscheduler-alert/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java]: Suggestion:
AlerServer should not has this send, put it to other class

Comment (elonlo) [reply dolphinscheduler-alert/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java]: yes, I want to remove the logic of regularly scanning the DB in the second step and then decouple the AlertServer class and the Send method.

**Comment by elonlo** [comment]
> We should use zk lock to let AlertServer in a distribute mode , right ?
> @qiaozhanwei @dailidong
my idea is to make AlertServer a basic jar package without distributed logic. which service needs to send an alert, add a dependency in the pom file and then call the alert api

**Comment by elonlo** [comment]
> We should use zk lock to let AlertServer in a distribute mode , right ?
> @qiaozhanwei @dailidong

my idea is to make AlertServer a basic jar package without distributed logic. which service needs to send an alert, add a dependency in the pom file and then call the alert api

File: dolphinscheduler-alert/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java (hunk 1 above)
Comment (Technoboy-) [reply dolphinscheduler-alert/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java]: > yes, I want to remove the logic of regularly scanning the DB in the second step and then decouple the AlertServer class and the Send method.

Merging your pr will affect all the developer.  Even though refactor can split into more steps, but every step should code nice .

Comment (elonlo) [reply dolphinscheduler-alert/src/main/java/org/apache/dolphinscheduler/alert/AlertServer.java]: hmm, I will abstract the Send function

**Pull request closed without merging**